### PR TITLE
Add linker variables for idle stack and heap

### PR DESCRIPTION
**Pull request by thapav**
This patch adds the following:
Idle stack start address (*s_idle_stack*): This will be used to place the idle stack at some particular location in the ram. Currently we are placing the idle stack at _ebss.
Heap start address (*_sext_heap_start*) (*_sint_heap_start*): The device might have internal and external ram. The internal ram will contain the data and bss, followed by the idle stack and heap. The external ram can also be used for heap. However, the starting portion of external ram might be used by vendor for storing some data. So, this new linker variable will indicate the start address of heap in the external ram.

### CODE REVIEW

File: os/arch/arm/src/armv7-m/up_vectors.c [hunk 1]
```diff
@@ -61,7 +61,7 @@
  * Preprocessor Definitions
  ************************************************************************************/
 
-#define IDLE_STACK      ((unsigned)&_ebss+CONFIG_IDLETHREAD_STACKSIZE-4)
+#define IDLE_STACK      ((unsigned)&_sidle_stack+CONFIG_IDLETHREAD_STACKSIZE-4)
```
[the remark this answers is not in the document]
Comment (thapav) [reply os/arch/arm/src/armv7-m/up_vectors.c:64]: Done

File: os/board/rtl8720e/src/component/soc/amebalite/fwlib/ram_km4/ameba_app_start.c [hunk 2]
```diff
@@ -43,15 +43,15 @@ extern u32 GlobalDebugEnable;
...
 extern unsigned int __StackStart;
 extern unsigned int __StackLimit;
-extern uint8_t __psram_heap_buffer_start__[];
+extern uint8_t _sext_heap_start[];
 extern unsigned int __PsramStackLimit;
 void NS_ENTRY BOOT_IMG3(void);
 void app_init_psram(void);
 
 #define IDLE_STACK ((uintptr_t)&__StackStart+CONFIG_IDLETHREAD_STACKSIZE-4)
 #define HEAP_BASE  ((uintptr_t)&__StackStart+CONFIG_IDLETHREAD_STACKSIZE)
```
[the remark this answers is not in the document]
Comment (thapav) [reply os/board/rtl8720e/src/component/soc/amebalite/fwlib/ram_km4/ameba_app_start.c:52]: I have removed for this and the file [os/board/rtl8721csm/src/component/soc/realtek/amebad/fwlib/ram_hp/rtl8721dhp_app_start.c]
Let us wait for changes from realtek. Will verify and add after that.

File: os/board/rtl8720e/src/component/soc/amebalite/fwlib/include/ameba_boot.h [hunk 3]
```diff
@@ -92,8 +92,7 @@ extern uint8_t __bdram_heap_buffer_start__[];
 extern uint8_t __bdram_heap_buffer_size__[];
 extern uint8_t __km4_sram_heap_extend_start__[];
 extern uint8_t __km4_sram_heap_extend_length__[];
```
[the remark this answers is not in the document]
Comment (thapav) [reply os/board/rtl8720e/src/component/soc/amebalite/fwlib/include/ameba_boot.h:94]: Done

File: build/configs/artik053/scripts/flash.ld [hunk 4]
```diff
@@ -119,6 +119,7 @@ SECTIONS
 		*(COMMON)
```
Comment (swasthik22) [inline build/configs/artik053/scripts/flash.ld:119]: Please change the commit message
 "the following unused linekr variables" -> "the following unused linker variables"

File: build/configs/rtl8721csm/scripts/rlx8721d_img2_is.ld [hunk 5]
```diff
@@ -233,6 +233,7 @@ SECTIONS
 		*(.bfsram.data*)
 		end = .;
 		_ebss = ABSOLUTE(.);
+		_sidle_stack = ABSOLUTE(.);
```
[the remark this answers is not in the document]
Comment (thapav) [reply build/configs/rtl8721csm/scripts/rlx8721d_img2_is.ld:236]: @sunghan-chang What do you think ?

[the remark this answers is not in the document]
Comment (sunghan-chang) [reply build/configs/rtl8721csm/scripts/rlx8721d_img2_is.ld:236]: @thapav I agree with @kishore-sn about adding new internal heap start. . But `_sidle_stack` is the allocation address, not top stack. So it should be `end of MSP_RAM_NS - CONFIG_IDLE_STACK_SIZE`. Could you confirm this?

[the remark this answers is not in the document]
Comment (thapav) [reply build/configs/rtl8721csm/scripts/rlx8721d_img2_is.ld:236]: It seems to work fine with _sidle_stack set as below :
_sidle_stack = ABSOLUTE(ORIGIN(MSP_RAM_NS) + LENGTH(MSP_RAM_NS)) - 1024;
_sidle_stack = ABSOLUTE(ORIGIN(KM4_MSP_RAM_NS) + LENGTH(KM4_MSP_RAM_NS)) - 1024;

File: build/configs/rtl8721csm/scripts/rlx8721d_img2_is.ld [hunk 6]
```diff
@@ -320,13 +322,16 @@ SECTIONS
...
+		/* Heap start address in external RAM */
+		_sext_heap_start = ABSOLUTE(.);
 	} > PSRAM_NS
 
 	.psram_heap.data : 
 	{
 		. = ALIGN (32);
 		*(.psram.heap*)
 	} > PSRAM_NS
+	_sidle_stack = ABSOLUTE(ORIGIN(MSP_RAM_NS) + LENGTH(MSP_RAM_NS)) - 1024;
```
Comment (sunghan-chang) [inline build/configs/rtl8721csm/scripts/rlx8721d_img2_is.ld:334]: `1024` should be changed to the CONFIG. `CONFIG_IDLETHREAD_STACKSIZE`.

File: build/configs/qemu/scripts/flash-sram.ld [hunk 7]
```diff
@@ -114,6 +114,9 @@ SECTIONS
 		*(.gnu.linkonce.b.*)
 		*(COMMON)
 		_ebss = ABSOLUTE(.);
+		_sidle_stack = ABSOLUTE(.);
+		/* Heap start address in internal RAM */
+		_sint_heap_start = ABSOLUTE(.);
```
Comment (sunghan-chang) [inline build/configs/qemu/scripts/flash-sram.ld:119]: It looks wrong. `_sidle_stack` and `_sint_heap_start` can't be the same. Because `_sint_heap_start` should be `_sidle_stack + CONFIG_IDLETHREAD_STACKSIZE`.

File: build/configs/stm32f407-disc1/scripts/kernel-space.ld [hunk 8]
```diff
@@ -107,6 +107,9 @@ SECTIONS
         *(.gnu.linkonce.b.*)
         *(COMMON)
         _ebss = ABSOLUTE(.);
+	_sidle_stack = ABSOLUTE(.);
+      /* Heap start address in internal RAM */
+      _sint_heap_start = ABSOLUTE(.);
```
Comment (sunghan-chang) [inline build/configs/stm32f407-disc1/scripts/kernel-space.ld:112]: Please use the same indentation. Other lines use a tab, not spaces.

**Review by sunghan-chang — COMMENTED** [review]
Could you remove `TizenRT` in the first commit title?
```
Add _sidle_stack variable to indicate the start of idle stack in TizenRT
```
Because we only support TizenRT, we can reduce a length of the title.

**Comment by sunghan-chang** [comment]
@thapav Once we confirm the change, please add this on the porting guide. Chipset vendor should do porting those new variable later.

File: build/configs/qemu/scripts/flash-sram.ld (hunk 7 above)
Comment (thapav) [reply build/configs/qemu/scripts/flash-sram.ld:119]: This is added in the start.c of chipsets as below:
#define HEAP_BASE ((uintptr_t)&_sint_heap_start + CONFIG_IDLETHREAD_STACKSIZE)

Comment (sunghan-chang) [reply build/configs/qemu/scripts/flash-sram.ld:119]: When someone sets the same address for both _sidle_stack and _sint_heap_start, then this calculation is correct. 
But
1. logically it's wrong, idle stack and heap start can't be the same.
2. If idle stack and internal heap is separate, then this calculation is wrong. At that case, the HEAP_BASE should be just _sint_heap_start.

File: build/configs/qemu/scripts/flash-sdram-qemu-memory-increased.ld [hunk 9]
```diff
@@ -114,6 +114,9 @@ SECTIONS
 		*(.gnu.linkonce.b.*)
 		*(COMMON)
 		_ebss = ABSOLUTE(.);
+		_sidle_stack = ABSOLUTE(.);
+		/* Heap start address in internal RAM */
+		_sint_heap_start = ABSOLUTE(.);
```
Comment (kishore-sn) [inline build/configs/qemu/scripts/flash-sdram-qemu-memory-increased.ld:119]: You can change these lines as below, so that you can use  directly in the code. 
```
_sidle_stack = ABSOLUTE(.);
. = CONFIG_IDLETHREAD_STACKSIZE ;
_sint_heap_start = ABSOLUTE(.);
```

**Comment by thapav** [comment]
@sunghan-chang @kishore-sn Updated the code with latest changes. 
Kindly review

File: build/configs/imxrt1050-evk/loadable_elf_apps/defconfig [hunk 10]
```diff
@@ -855,7 +855,9 @@ CONFIG_MEMSET_OPTSPEED=y
 # CONFIG_ARCH_BZERO is not set
 # CONFIG_LIB_ENVPATH is not set
 # CONFIG_LIB_HASHMAP is not set
-
+CONFIG_EXAMPLES_WIFIMANAGER_TEST=y
+CONFIG_WIFIMANAGER_TEST_TRIAL=1
+CONFIG_EXAMPLES_WIFIMANAGER_AP_LIST_ITEMS_COUNT=10
```
Comment (kishore-sn) [inline build/configs/imxrt1050-evk/loadable_elf_apps/defconfig:860]: I think these changes are not related to current PR. Can we revert this file?

File: build/configs/rtl8721csm/scripts/rlx8721d_img2_is.ld [hunk 11]
```diff
@@ -233,6 +233,9 @@ SECTIONS
 		*(.bfsram.data*)
 		end = .;
 		_ebss = ABSOLUTE(.);
+		. = . + CONFIG_IDLETHREAD_STACKSIZE ;
```
Comment (kishore-sn) [inline build/configs/rtl8721csm/scripts/rlx8721d_img2_is.ld:236]: Since we are not placing the idle stack here, we dont have to allocate space for it here.

File: build/configs/rtl8721csm/scripts/rlx8721d_img2_ns.ld [hunk 12]
```diff
@@ -306,6 +306,9 @@ SECTIONS
 		*(.bfsram.data*)
 		end = .;
 		_ebss = ABSOLUTE(.);
+		. = . + CONFIG_IDLETHREAD_STACKSIZE ;
```
Comment (kishore-sn) [inline build/configs/rtl8721csm/scripts/rlx8721d_img2_ns.ld:309]: Please remove this allocation for idle stack.

File: os/arch/arm/src/common/up_internal.h [hunk 13]
```diff
@@ -269,6 +269,7 @@ EXTERN uint32_t _sdata;			/* Start of .data */
 EXTERN uint32_t _edata;			/* End+1 of .data */
 EXTERN uint32_t _sbss;			/* Start of .bss */
 EXTERN uint32_t _ebss;			/* End+1 of .bss */
+EXTERN uint32_t _sidle_stack;		/* Start of idle stack */
```
Comment (kishore-sn) [inline os/arch/arm/src/common/up_internal.h:272]: How about adding the variables for internal and external heap also here?

Comment (thapav) [reply os/arch/arm/src/common/up_internal.h:272]: Correct, updated them

File: os/board/rtl8721csm/src/component/soc/realtek/amebad/fwlib/ram_hp/rtl8721dhp_app_start.c [hunk 14]
```diff
@@ -44,13 +44,15 @@ void app_init_psram(void);
...
+extern unsigned int _sext_heap_start;
 extern unsigned int __StackLimit;
 extern unsigned int __PsramStackLimit;
-#define IDLE_STACK ((uintptr_t)&_ebss+CONFIG_IDLETHREAD_STACKSIZE-4)
-#define HEAP_BASE  ((uintptr_t)&_ebss+CONFIG_IDLETHREAD_STACKSIZE)
+#define IDLE_STACK ((uintptr_t)&_sidle_stack + CONFIG_IDLETHREAD_STACKSIZE - 4)
+#define HEAP_BASE  ((uintptr_t)&_sint_heap_start)
 #define HEAP_LIMIT ((uintptr_t)&__StackLimit)
-#define PSRAM_HEAP_BASE ((uintptr_t)&__psram_bss_end__[0])
+#define PSRAM_HEAP_BASE ((uintptr_t)&_sext_heap_start)
```
Comment (kishore-sn) [inline os/board/rtl8721csm/src/component/soc/realtek/amebad/fwlib/ram_hp/rtl8721dhp_app_start.c:55]: Please update the value of `g_idle_topstack` in `os/arch/arm/src/amebad/amebad_start.c`

```
#define MSP_RAM_HP_NS           0x10004FFC
const uintptr_t g_idle_topstack = MSP_RAM_HP_NS;
```

Comment (thapav) [reply os/board/rtl8721csm/src/component/soc/realtek/amebad/fwlib/ram_hp/rtl8721dhp_app_start.c:55]: Updated the g_idle_topstack values. Thank you

**Review by kishore-sn — DISMISSED** [review]
Looks good.

File: docs/HowToAddnewBoard.md [hunk 15]
```diff
@@ -211,3 +211,15 @@ This is moved to .config at executing configure.sh and used to build TizenRT.
...
+Linker script has memory configuration information.  
+file path : *build/configs/<BOARD_NAME>/scripts/<SCRIPTS_NAME>.ld* 
+
+In addition to the SECTION information, add the following variables as well to mark the following:  
+1. `_sbss`: Start of .bss section
+2. `_ebss`: End+1 of .bss section
+3. `_sidle_stack`: Start of idle stack
+4. `__sint_heap_start`: Start of heap in internal RAM region
+5. `__sext_heap_start`: Start of heap in external RAM region
+
```
Comment (Tirtha-Samsung) [inline docs/HowToAddnewBoard.md:225]: Looks good to me

**Review by Tirtha-Samsung — COMMENTED** [review]
Looks good to me

File: docs/HowToAddnewBoard.md [hunk 16]
```diff
@@ -211,3 +211,15 @@ This is moved to .config at executing configure.sh and used to build TizenRT.
...
+Linker script has memory configuration information.  
+file path : *build/configs/<BOARD_NAME>/scripts/<SCRIPTS_NAME>.ld* 
+
+In addition to the SECTION information, add the following variables as well to mark the following:  
+1. `_sbss`: Start of .bss section
+2. `_ebss`: End+1 of .bss section
+3. `_sidle_stack`: Start of idle stack
+4. `__sint_heap_start`: Start of heap in internal RAM region
+5. `__sext_heap_start`: Start of heap in external RAM region
```
Comment (sunghan-chang) [inline docs/HowToAddnewBoard.md:224]: They need some guidance about where chipset vendor should locate them.
Let's add two examples for artik and rtl8720e.

**Comment by sunghan-chang** [comment]
Could you check idle coloration on rtl8721csm?

**Comment by thapav** [comment]
> Could you check idle coloration on rtl8721csm?

This was verified

**Review by sunghan-chang — APPROVED** [review]
